code/modules/power/turbine/turbine.dm

```diff
@@ -0,0 +1,557 @@
+		return
+
+	balloon_alert(user, "already installed")
+	return
+
+/obj/machinery/power/turbine/core_rotor
+	name = "core rotor"
+	desc = "The middle part of a turbine generator, contains the rotor and the main computer."
+	icon = 'icons/obj/turbine/turbine.dmi'
+	icon_state = "core_rotor"
+
+	circuit = /obj/item/circuitboard/machine/turbine_rotor
+
+	gas_theoretical_volume = 3000
+
+	part_path = /obj/item/turbine_parts/rotor
+
+	///ID to easily connect the main part of the turbine to the computer
+	var/mapping_id
+
+	///Reference to the compressor
+	var/obj/machinery/power/turbine/inlet_compressor/compressor
+	///Reference to the turbine
+	var/obj/machinery/power/turbine/turbine_outlet/turbine
+
+	///Reference to the input turf
+	var/turf/open/input_turf
+	///Reference to the output turf
+	var/turf/open/output_turf
+
+	///Efficiency of the part installed in the compressor (from 0.25 to 0.85)
+	var/compressor_part_efficiency = 0.25
+	///Efficiency of the part installed in the rotor (from 0.25 to 0.85)
+	var/rotor_part_efficiency = 0.25
+	///Efficiency of the part installed in the turbine (from 0.85 to 0.895) - don't go higher than that, it can cause #nan errors
+	var/stator_part_efficiency = 0.85
+
+	///Rotation per minute the machine is doing
+	var/rpm
+	///Amount of power the machine is producing
+	var/produced_energy
+
+	///First stage of the moving gasmix - compression from 2500 L to 1000 L - heat up
+	var/datum/gas_mixture/compressor_mixture
+	///Second stage of the moving gasmix - expansion from 1000 L to 3000 L - first cool down
+	var/datum/gas_mixture/rotor_mixture
+	///Third stage of the moving gasmix - expansion from 3000 L to 6000 L - second cool down
+	var/datum/gas_mixture/turbine_mixture
+
+	///Check to see if all parts are connected to the core
+	var/all_parts_connected = FALSE
+	///If the machine was completed before reopening it, try to remake it
+	var/was_complete = FALSE
+
+	///Max rmp that the installed parts can handle, limits the rpms
+	var/max_allowed_rpm = 0
+	///Max temperature that the installed parts can handle, unlimited and causes damage to the machine
+	var/max_allowed_temperature = 0
+
+	///Amount of damage the machine has received
+	var/damage = 0
+	///Used to calculate the max damage received per tick and if the alarm should be called
+	var/damage_archived = 0
+
+	///Our internal radio
+	var/obj/item/radio/radio
+	///The key our internal radio uses
+	var/radio_key = /obj/item/encryptionkey/headset_eng
+	///The engineering channel
+	var/engineering_channel = "Engineering"
+
+	COOLDOWN_DECLARE(turbine_damage_alert)
+
+/obj/machinery/power/turbine/core_rotor/Initialize(mapload)
+	. = ..()
+	radio = new(src)
+	radio.keyslot = new radio_key
+	radio.set_listening(FALSE)
+	radio.recalculateChannels()
+
+/obj/machinery/power/turbine/core_rotor/LateInitialize()
+	. = ..()
+	activate_parts()
+
+/obj/machinery/power/turbine/core_rotor/Destroy(mob/user)
+	deactivate_parts(user)
+	QDEL_NULL(radio)
+	return ..()
```

Comment:
How does a user get here exactly?

Reply:
By breaking the machine or deconstruct it

Reply:
how are you passing that into Destroy?